trie/proof.go

```diff
@@ -472,12 +472,17 @@ func VerifyRangeProof(rootHash common.Hash, firstKey []byte, lastKey []byte, key
 	if len(keys) != len(values) {
 		return false, fmt.Errorf("inconsistent proof data, keys: %d, values: %d", len(keys), len(values))
 	}
-	// Ensure the received batch is monotonic increasing.
+	// Ensure the received batch is monotonic increasing and contains no deletions
 	for i := 0; i < len(keys)-1; i++ {
 		if bytes.Compare(keys[i], keys[i+1]) >= 0 {
 			return false, errors.New("range is not monotonically increasing")
 		}
 	}
+	for _, value := range values {
+		if len(value) == 0 {
+			return false, errors.New("range contains deletion")
+		}
+	}
```

Comment:
`len(keys) == len(values)`, so you could use the same loop, with a slight modification
```suggestion
	// Ensure the received batch is monotonic increasing and contains no deletions
	for i := 0; i < len(keys); i++ {
		if i > 0 && bytes.Compare(keys[i-1], keys[i]) >= 0 {
			return false, errors.New("range is not monotonically increasing")
		}
		if len(values[i]) == 0 {
			return false, errors.New("range contains deletion")
		}
	}
```
Feel free to ignore

Reply:
👍

Reply:
It did occur to me, but the off by one and extra clause check seemed wonky. The code is cleaner this way, we're not saving anything by optimizing out an extra addition per array item. Unless you guys feel very strongly, I'd leave it as is.